### PR TITLE
Graceful shutdown. 

### PR DESCRIPTION
**Pull request by kflansburg**
I wanted to invite comments on what I have been working on:

* Catch sigterm and certain errors and trigger graceful shutdown.
* Graceful shutdown cordons node and evicts pods, before deregistering it with the apiserver.
* Refactor kubelet futures for readability.
* Rework the combinators used for kubelet futures to enable graceful error handling.

This seems to do a good job of allowing Kubernetes to reschedule pods, rather than leaving zombie pods, and it gives the provider an opportunity to properly clean up pod resources as they are deleted. I'm not totally sure if all of this is desired behavior so I wanted to get some feedback before continuing.

### CODE REVIEW

**Comment by thomastaylor312** [comment]
I think we are completely open to this @kflansburg. The reason we hadn't focused on it was that the WASM providers don't leave zombie pods when they restart as they are all running within the krustlet process. I'll go leave comments in the code (if I have any) on the approaches you are taking.

Also, going to mark this as a draft for now just so we know when it is ready for a final review by pulling it out of a draft state

File: crates/kubelet/src/kubelet.rs [hunk 1]
```diff
@@ -54,98 +54,77 @@ impl<T: 'static + Provider + Sync + Send> Kubelet<T> {
...
         let client = kube::Client::new(self.kube_config.clone());
+
         // Create the node. If it already exists, "adopt" the node definition
         create_node(&client, &self.config, self.provider.clone()).await;
 
-        // Get the node name for use in the update loop
-        let node_name = self.config.node_name.clone();
+        // Flag to indicate graceful shutdown has started.
+        let signal = Arc::new(AtomicBool::new(false));
+        signal_hook::flag::register(signal_hook::SIGINT, Arc::clone(&signal))?;
```
Comment (thomastaylor312) [inline crates/kubelet/src/kubelet.rs:63]: Since we are alerting only on SIGINT, we should probably just leverage `tokio::signal::ctrl_c`

File: crates/kubelet/src/kubelet.rs [hunk 2]
```diff
@@ -162,6 +141,107 @@ impl<P> Clone for Kubelet<P> {
...
+) -> anyhow::Result<()> {
+    let node_selector = format!("spec.nodeName={}", node_name);
+    let params = ListParams {
+        field_selector: Some(node_selector),
+        ..Default::default()
+    };
+    let api = Api::<KubePod>::all(client);
+    let informer = Informer::new(api).params(params);
+    loop {
+        // TODO Should these results be ignored if Err? It seems reasonable to just keep trying,
```
Comment (thomastaylor312) [inline crates/kubelet/src/kubelet.rs:158]: I agree here. Just log the error and let it keep trying so this can handle the graceful shutdown

**Comment by kflansburg** [comment]
Sounds good, while there may be nothing running after the WASM provider shuts down, the node remains registered with kube-apiserver, and pods remain with a "Running" state. Maybe zombie isn't the best term for this.

File: crates/kubelet/src/node.rs [hunk 3]
```diff
@@ -166,6 +167,124 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
...
+        field_selector: Some(node_selector),
+        ..Default::default()
+    };
+    let kube::api::ObjectList { items: pods, .. } = pod_client.list(&params).await?;
+
+    warn!("Evicting {} pods.", pods.len());
+
+    // TODO: Filter mirror pods,daemonsets, kube-system?
+    for pod in &pods {
+        if let Some(name) = pod.metadata.as_ref().and_then(|meta| meta.name.as_ref()) {
```
Comment (thomastaylor312) [inline crates/kubelet/src/node.rs:211]: You can bring the `Meta` trait into scope to avoid all of this unwrapping. Also, it may be worth it to convert all of these to our internal `Pod` wrapper to get convenient accessors on this data

File: crates/kubelet/src/node.rs [hunk 4]
```diff
@@ -166,6 +167,124 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
...
+
+async fn evict_pod(client: &kube::Client, name: &str, namespace: &str) -> anyhow::Result<()> {
+    let ns_client: Api<Pod> = Api::namespaced(client.clone(), namespace);
+    let lp = ListParams::default()
+        .fields(&format!("metadata.name={}", name))
+        .fields(&format!("metadata.namespace={}", namespace));
+
+    // The delete call may return a "pending" response, we must watch for the actual delete event.
+    // TODO Timeout?
+    let mut stream = ns_client.watch(&lp, "0").await?.boxed();
```
Comment (thomastaylor312) [inline crates/kubelet/src/node.rs:239]: I'm not sure if it matters much in this PR, but I thought I'd bring it up: We might want to check if there is any performance overhead for spinning up a separate watch for each pod. If there are 100s of pods on the node, I don't know what the API Server does with the list params and what the overhead is for opening a stream for each individual pod. If it is a problem, we might want to see if there is a good way to share the same watch

File: crates/kubelet/src/kubelet.rs [hunk 5]
```diff
@@ -162,6 +141,107 @@ impl<P> Clone for Kubelet<P> {
...
+    signal: Arc<AtomicBool>,
+    client: kube::Client,
+    node_name: String,
+) -> anyhow::Result<()> {
+    let duration = std::time::Duration::from_millis(100);
+    loop {
+        if signal.load(Ordering::Relaxed) {
+            warn!("Signal caught.");
+            drain_node(&client, &node_name).await?;
+            delete_node(&client, &node_name).await?;
```
Comment (thomastaylor312) [inline crates/kubelet/src/kubelet.rs:196]: If I remember correctly, the normal kubelet doesn't deregister itself from the API. I think cordoning would be ok, but deleting probably isn't needed

File: crates/kubelet/src/kubelet.rs (hunk 1 above)
Comment (kflansburg) [reply crates/kubelet/src/kubelet.rs:63]: I've used ctrl_c in the past, but we may wish to support other signals such as SIGTERM, if, for instance, we wish to run Krustlet inside of a container.

File: crates/kubelet/src/kubelet.rs (hunk 2 above)
Comment (kflansburg) [reply crates/kubelet/src/kubelet.rs:158]: Will do.

File: crates/kubelet/src/node.rs (hunk 3 above)
Comment (kflansburg) [reply crates/kubelet/src/node.rs:211]: Yeah, I was on the fence about using the `Pod` wrapper, I will try it out.

File: crates/kubelet/src/node.rs (hunk 4 above)
Comment (kflansburg) [reply crates/kubelet/src/node.rs:239]: True, I'm not sure how `kubectl` [does it](https://kubernetes.io/images/docs/kubectl_drain.svg), I don't think its very sophisticated. I think it would be possible to do with a single `Stream` so I can give that a try.

File: crates/kubelet/src/kubelet.rs (hunk 5 above)
Comment (kflansburg) [reply crates/kubelet/src/kubelet.rs:196]: Hmmm, ok. I'm fine with this but I think then I would like to look at the re-registration functionality because I have been having issues with that.

File: crates/kubelet/src/kubelet.rs (hunk 1 above)
Comment (thomastaylor312) [reply crates/kubelet/src/kubelet.rs:63]: Yeah, I figure we could, although windows support for that is a bit hairy.

Also, we may want to use the `tokio-support` feature to convert this into a `Stream`

File: crates/kubelet/src/node.rs (hunk 4 above)
Comment (thomastaylor312) [reply crates/kubelet/src/node.rs:239]: If it becomes too complex, don't sweat it, we can just leave a comment and revisit later

File: crates/kubelet/src/kubelet.rs (hunk 5 above)
Comment (thomastaylor312) [reply crates/kubelet/src/kubelet.rs:196]: So right now the re-registration is a hack because we don't have TLS bootstrapping and can't update the node status directly. Another community member is almost done with that (we are waiting to cut 0.2 for it), so this should be greatly improved soon-ish

File: crates/kubelet/src/kubelet.rs (hunk 1 above)
Comment (kflansburg) [reply crates/kubelet/src/kubelet.rs:63]: Ah, I see.

File: crates/kubelet/src/kubelet.rs (hunk 5 above)
Comment (kflansburg) [reply crates/kubelet/src/kubelet.rs:196]: Ok, sounds good.

File: crates/kubelet/src/node.rs [hunk 6]
```diff
@@ -166,6 +168,120 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
...
+    let params = ListParams {
+        field_selector: Some(node_selector),
+        ..Default::default()
+    };
+    let kube::api::ObjectList { items: pods, .. } = pod_client.list(&params).await?;
+
+    let lp = ListParams::default().fields(&format!("spec.nodeName={}", node_name));
+
+    // The delete call may return a "pending" response, we must watch for the actual delete event.
+    // TODO I dont know what the version number does here.
```
Comment (thomastaylor312) [inline crates/kubelet/src/node.rs:211]: This is the resource version of the object in Kubernetes. From the docs:
> When specified with a watch call, shows changes that occur after that particular version of a resource. Defaults to changes from the beginning of history. When specified for list: - if unset, then the result is returned from remote storage based on quorum-read flag; - if it's 0, then we simply return what we currently have in cache, no guarantee; - if set to non zero, then the result is at least as fresh as given rv.

File: crates/kubelet/src/node.rs [hunk 7]
```diff
@@ -166,6 +168,120 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
...
+    let mut stream = pod_client.watch(&lp, "0").await?.boxed();
+
+    warn!("Evicting {} pods.", pods.len());
+
+    // TODO: Filter mirror pods,daemonsets, kube-system?
```
Comment (thomastaylor312) [inline crates/kubelet/src/node.rs:216]: For feature parity with what `kubectl drain` does, it should _technically_ error if it can't remove these kinds of pods (except for mirror, those are ignored). However, we probably want the `--ignore-daemonset` equivalent and don't care if we error here anyway.

tl;dr: Ignore mirror and daemonsets. Kube-system should be evicted if they are not daemonsets

**Comment by thomastaylor312** [comment]
@kflansburg This is looking close! Feel free to mark as ready for review when you'd like me to review

File: crates/kubelet/src/node.rs (hunk 7 above)
Comment (kflansburg) [reply crates/kubelet/src/node.rs:216]: Ok, I will give this a shot. In my testing, kube-proxy gets rescheduled onto the node and can be left as a zombie. Do you think it would be reasonable to have `pod_informer` drop add pod events after shutdown has started to prevent this scheduling?

**Comment by kflansburg** [comment]
Ok, I have added filtering for static pods and daemon sets. For static pods, I still patch the status of the pod as terminated. I think this is most informative to consumers of the kube-api and it causes the pod to be rescheduled on the node when it is restarted.

File: crates/kubelet/src/kubelet.rs [hunk 8]
```diff
@@ -162,6 +142,120 @@ impl<P> Clone for Kubelet<P> {
     }
 }
 
+/// Awaits SIGINT and sets graceful shutdown flag if detected.
+async fn start_signal_task(signal: Arc<AtomicBool>) -> anyhow::Result<()> {
+    ctrl_c().await?;
+    warn!("Caught keyboard interrupt.");
+    signal.store(true, Ordering::Relaxed);
```
Comment (rylev) [inline crates/kubelet/src/kubelet.rs:149]: nit: it'd be good to have a comment here on ordering and why `Relaxed` is the right one.

File: crates/kubelet/src/node.rs [hunk 9]
```diff
@@ -166,6 +169,142 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
     info!("Successfully created node '{}'", &config.node_name);
 }
 
+/// Fetch the uid of a node by name.
+pub async fn node_uid(client: &kube::Client, node_name: &str) -> anyhow::Result<String> {
+    let node_client: Api<KubeNode> = Api::all(client.clone());
+    match retry!(node_client.get(node_name).await, times: 4, log_error: |e| error!("Failed to get node to cordon: {:?}", e))
+    {
+        Ok(node) => match node.metadata.and_then(|meta| meta.uid) {
```
Comment (rylev) [inline crates/kubelet/src/node.rs:177]: nit can the `node` be matched on inline inside of the `Ok`? 
`Ok(KubeNode { metadata: Some(Metadata { uid: Some(uid) }`

File: crates/kubelet/src/node.rs [hunk 10]
```diff
@@ -166,6 +169,142 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
...
+    // The delete call may return a "pending" response, we must watch for the actual delete event.
+    let mut stream = pod_client.watch(&lp, "0").await?.boxed();
+
+    warn!("Evicting {} pods.", pods.len());
+
+    for pod in pods {
+        let pod = Pod::new(pod);
+        if pod.is_daemonset() {
+            warn!("Skipping eviction of DaemonSet '{}'", pod.name());
+            continue;
```
Comment (rylev) [inline crates/kubelet/src/node.rs:220]: nit: why `if` with `continue` instead of `if { } else if {}`?

File: crates/kubelet/src/node.rs [hunk 11]
```diff
@@ -166,6 +169,142 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
...
+                    if (name == pod.name()) & (namespace == pod.namespace()) {
+                        warn!("Pod '{}' evicted.", name);
+                        break;
+                    }
+                }
+                _ => (),
+            }
+        }
+    } else {
+        warn!("Pod '{}' evicted.", name);
```
Comment (rylev) [inline crates/kubelet/src/node.rs:287]: Should all these logs be `debug` not `warn`?

File: crates/kubelet/src/node.rs [hunk 12]
```diff
@@ -166,6 +169,142 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
...
+    let response = ns_client.delete(name, &params).await?;
+
+    if response.is_left() {
+        // TODO Timeout?
+        warn!("Waiting for pod '{}' eviction.", name);
+        while let Some(event) = stream.try_next().await? {
+            match event {
+                kube::api::WatchEvent::Deleted(s) => {
+                    let pod = Pod::new(s);
+                    if (name == pod.name()) & (namespace == pod.namespace()) {
```
Comment (rylev) [inline crates/kubelet/src/node.rs:278]: Did you mean to do a bitwise `and` here?

File: crates/kubelet/src/node.rs [hunk 13]
```diff
@@ -166,6 +169,142 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
...
+        }
+    } else {
+        warn!("Pod '{}' evicted.", name);
+    }
+    Ok(())
+}
+
+/// Mark this node as unschedulable.
+pub async fn cordon_node(client: &kube::Client, node_name: &str) -> anyhow::Result<()> {
+    warn!("Cordoning node.");
```
Comment (rylev) [inline crates/kubelet/src/node.rs:294]: nit: log name of node

File: crates/kubelet/src/pod.rs [hunk 14]
```diff
@@ -84,6 +84,26 @@ impl Pod {
...
+    /// dont know how reliable this is.
+    pub fn is_static(&self) -> bool {
+        self.0.meta().owner_references.is_none()
+    }
+
+    /// Indicate if this pod is part of a Daemonset
+    pub fn is_daemonset(&self) -> bool {
+        if let Some(owners) = &self.0.meta().owner_references {
+            for owner in owners {
+                match owner.kind.as_ref() {
```
Comment (rylev) [inline crates/kubelet/src/pod.rs:98]: nit: use `if let` instead of match here

Comment (kflansburg) [reply crates/kubelet/src/pod.rs:98]: Just curious, what makes one better than the other?

File: crates/kubelet/src/node.rs (hunk 12 above)
Comment (kflansburg) [reply crates/kubelet/src/node.rs:278]: I don't follow. Isn't that a logical and of two `bool`s?

File: crates/kubelet/src/node.rs (hunk 10 above)
Comment (kflansburg) [reply crates/kubelet/src/node.rs:220]: It started out as just a one line check and was more readable that way. I will change it.

File: crates/kubelet/src/kubelet.rs (hunk 8 above)
Comment (kflansburg) [reply crates/kubelet/src/kubelet.rs:149]: Well, my reasoning isn't very specific. We aren't trying to synchronize anything here. Once the flag is set, other tasks discover this with best effort and we don't really need any hard requirements on when this happens.

Comment (kflansburg) [reply crates/kubelet/src/kubelet.rs:149]: I think this could also be implemented with a Mutex<bool> or something and ordering would be a non-issue.

File: crates/kubelet/src/node.rs (hunk 7 above)
Comment (thomastaylor312) [reply crates/kubelet/src/node.rs]: If you can find a nice way to let the pod informer know, I think that is a good idea

File: crates/kubelet/src/node.rs (hunk 11 above)
Comment (thomastaylor312) [reply crates/kubelet/src/node.rs:287]: I think `info` is the right level here

File: crates/kubelet/src/node.rs (hunk 12 above)
Comment (thomastaylor312) [reply crates/kubelet/src/node.rs:278]: This is a bitwise operation rather than a boolean operation `&` vs `&&`

Comment (kflansburg) [reply crates/kubelet/src/node.rs:278]: I don't believe this distinction applies to `bool`. If you want to remove the parens and compare the expressions logically, I can do that.

File: crates/kubelet/src/node.rs (hunk 11 above)
Comment (kflansburg) [reply crates/kubelet/src/node.rs:287]: ok, will do

File: crates/kubelet/src/node.rs (hunk 12 above)
Comment (thomastaylor312) [reply crates/kubelet/src/node.rs:278]: I think the end result is the same, but the bitwise operator might have other performance/side effects I don't know about. I'll leave it to @rylev to clarify

File: crates/kubelet/src/pod.rs (hunk 14 above)
Comment (rylev) [reply crates/kubelet/src/pod.rs]: It’s just a bit more terse when only one pattern is being handled. It’s the more idiomatic thing to do.

Comment (rylev) [reply crates/kubelet/src/pod.rs]: It’s just a bit more terse when only one pattern is being handled. It’s the more idiomatic thing to do.

Comment (rylev) [reply crates/kubelet/src/pod.rs]: It’s just a bit more terse when only one pattern is being handled. It’s the more idiomatic thing to do.

File: crates/kubelet/src/node.rs (hunk 12 above)
Comment (rylev) [reply crates/kubelet/src/node.rs]: I don’t believe it makes a difference runtime wise. It’s just very surprising to see bite wise and. Seems clever when it doesn’t have to be

**Comment by thomastaylor312** [comment]
@kflansburg You ready for final review on this?

**Comment by kflansburg** [comment]
@thomastaylor312 Yep, ready for review.

File: crates/kubelet/src/node.rs [hunk 15]
```diff
@@ -166,6 +170,140 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
...
+    namespace: &str,
+    stream: &mut PodStream,
+) -> anyhow::Result<()> {
+    let ns_client: Api<KubePod> = Api::namespaced(client.clone(), namespace);
+    info!("Evicting namespace '{}' pod '{}'", namespace, name);
+    let params = Default::default();
+    let response = ns_client.delete(name, &params).await?;
+
+    if response.is_left() {
+        // TODO Timeout?
```
Comment (thomastaylor312) [inline crates/kubelet/src/node.rs:274]: If you want to, this can be done with the `retry!` macro in this file. Not going to block on it though

**Comment by bacongobbler** [comment]
Thanks @kflansburg! Let's get this in and follow up with future PRs if you'd like to address @thomastaylor312's comment. Great work here.
